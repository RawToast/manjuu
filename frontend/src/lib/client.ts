import { z } from 'zod'

const AuthoritySchema = z.object({
  name: z.string().min(1),
  id: z.number().int().min(0),
  establishments: z.number().int().min(0)
})

export type Authority = z.infer<typeof AuthoritySchema>

export interface StandardRatings {
  five: number
  four: number
  three: number
  two: number
  one: number
  zero: number
  exempt: number
}

export interface ScottishRatings {
  pass: number
  improvementRequired: number
  exempt: number
}

export interface Ratings {
  Scottish: ScottishRatings | null
  Standard: StandardRatings | null
}

export interface AuthoritySummary {
  name: string
  ratings: Ratings
  establishments: number
  url: string
}

export async function fetchAuthorities(): Promise<Authority[]> {
  const response = await fetch(`${import.meta.env.VITE_API_BASE_URL}/authority`, {
    method: 'GET',
    headers: {
      accept: 'application/json'
    }
  })

  const json = await response.json()
  const result = z.array(AuthoritySchema).parse(json)
  return result
}

export async function fetchAuthorityStats(id: string): Promise<AuthoritySummary> {
  const response = await fetch(`${import.meta.env.VITE_API_BASE_URL}/authority/${id}`, {
    method: 'GET',
    headers: {
      accept: 'application/json'
    }
  })

  const json = await response.json()
  const result = json as AuthoritySummary
  return result
}

export interface ApiUser {
  id: string
  org_id: string
  is_superuser: boolean
}
