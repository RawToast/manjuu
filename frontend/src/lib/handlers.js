// src/mocks/handlers.js
import { http, HttpResponse } from 'msw'

const fakeAuthorities = [
  {
    name: 'Test Authority',
    establishments: 123,
    url: 'https://www.norwich.gov.uk/site/custom_scripts/eh/foodratings.html',
    ratings: {
      five: 11,
      four: 24,
      three: 38,
      two: 41,
      one: 7,
      zero: 0,
      exempt: 4
    }
  },
  {
    name: 'Fake Edinburgh',
    establishments: 572,
    url: 'https://www.norwich.gov.uk/site/custom_scripts/eh/foodratings.html',
    ratings: {
      pass: 398,
      improvement: 164,
      exempt: 10
    }
  }
]

export const handlers = [
  http.get('authority/:id', ({ params }) => {
    const authorityId = params.id
    const authIsOdd = authorityId % 2 == 1

    if (authIsOdd) {
      return HttpResponse.json(fakeAuthorities[0])
    } else {
      return HttpResponse.json(fakeAuthorities[1])
    }
  })
]
