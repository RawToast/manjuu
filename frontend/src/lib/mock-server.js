import { http, HttpResponse } from 'msw'
import { setupServer } from 'msw/node'

export const fakeAuthorities = [
  {
    id: 1,
    name: 'Fake Newcastle',
    establishments: 123,
    url: 'https://www.norwich.gov.uk/site/custom_scripts/eh/foodratings.html',
    ratings: {
      five: 11,
      four: 24,
      three: 38,
      two: 41,
      one: 7,
      zero: 0,
      exempt: 4
    }
  },
  {
    id: 2,
    name: 'Fake Edinburgh',
    establishments: 572,
    url: 'https://www.norwich.gov.uk/site/custom_scripts/eh/foodratings.html',
    ratings: {
      pass: 398,
      improvement: 164,
      exempt: 10
    }
  }
]

const handlers = [
  http.get('http://localhost:8080/authority/:id', ({ params }) => {
    const authorityId = params.id
    const auth = fakeAuthorities.filter((auth) => auth.id === authorityId)
    if (auth.length > 0) {
      return HttpResponse.json(auth[0])
    } else {
      return HttpResponse.json({ error: 'Authority not found' }, 404)
    }
  }),
  http.get('http://localhost:8080/authority', () => {
    return HttpResponse.json(fakeAuthorities)
  })
]

export const server = setupServer(...handlers)
