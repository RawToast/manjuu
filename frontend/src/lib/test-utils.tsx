import { QueryClient } from '@tanstack/react-query'
import {
  Outlet,
  createRoute,
  createMemoryHistory,
  createRouter,
  createRootRouteWithContext
} from '@tanstack/react-router'

export function createTestRouter(component: () => React.ReactNode, queryClient: QueryClient) {
  const rootRoute = createRootRouteWithContext<{ queryClient: QueryClient }>()({
    component: Outlet
  })

  const componentRoute = createRoute({
    getParentRoute: () => rootRoute,
    path: '/',
    component
  })

  const router = createRouter({
    routeTree: rootRoute.addChildren([componentRoute]),
    history: createMemoryHistory(),
    context: { queryClient }
  })

  return router
}
