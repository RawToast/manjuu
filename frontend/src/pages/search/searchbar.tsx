import {
  BuildingOfficeIcon,
  BuildingOffice2Icon,
  HomeIcon,
  HomeModernIcon
} from '@heroicons/react/24/outline'
import {
  Command,
  CommandEmpty,
  CommandGroup,
  CommandInput,
  CommandItem,
  CommandList,
  CommandSeparator
} from '@/components/ui/command'
import { Authority, fetchAuthorities } from '@/lib/client'
import React from 'react'
import { useQuery } from '@tanstack/react-query'

export function SearchBar() {
  const { status, data, error } = useQuery({ queryKey: ['todos'], queryFn: fetchAuthorities })

  const [searchText, setSearchText] = React.useState('')

  if (status === 'error') {
    return <span>Error: {error.message}</span>
  }

  return (
    <Command className='rounded-lg border shadow-md'>
      <CommandInput
        placeholder='Type a command or search...'
        value={searchText}
        onValueChange={setSearchText}
      />
      {searchText.length == 0 ? (
        <></>
      ) : (
        <CommandList>
          <CommandEmpty>No results found.</CommandEmpty>

          <CommandGroup heading='Authorities'>
            {status === 'pending' ? (
              <CommandEmpty>Loading...</CommandEmpty>
            ) : (
              data.map((authority) => {
                return <AuthorityItem key={authority.id} authority={authority}></AuthorityItem>
              })
            )}
          </CommandGroup>

          <CommandSeparator />
          <CommandGroup heading='Recent'>
            <AuthorityItem
              authority={{ id: 10, name: 'Matlock', establishments: 65 }}
            ></AuthorityItem>
          </CommandGroup>
        </CommandList>
      )}
    </Command>
  )
}

function AuthorityItem({ authority }: { authority: Authority }) {
  return (
    <CommandItem key={authority.id}>
      {authority.establishments > 3000 ? (
        <BuildingOffice2Icon className='mr-2 h-4 w-4' />
      ) : authority.establishments > 2000 ? (
        <BuildingOfficeIcon className='mr-2 h-4 w-4' />
      ) : authority.establishments > 1000 ? (
        <HomeModernIcon className='mr-2 h-4 w-4' />
      ) : (
        <HomeIcon className='mr-2 h-4 w-4' />
      )}
      <span>{authority.name}</span>
      {/* <CommandShortcut>âŒ˜1</CommandShortcut> */}
    </CommandItem>
  )
}
