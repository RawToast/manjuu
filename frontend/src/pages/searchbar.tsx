import {
  BuildingOfficeIcon,
  BuildingOffice2Icon,
  HomeIcon,
  HomeModernIcon
} from '@heroicons/react/24/outline'
import {
  Command,
  CommandEmpty,
  CommandGroup,
  CommandInput,
  CommandItem,
  CommandList,
  CommandSeparator
} from '@/components/ui/command'
import { Authority, fetchAuthorities } from '@/lib/client'
import { useState } from 'react'
import { useQuery } from '@tanstack/react-query'
import { Link } from '@tanstack/react-router'

export function SearchBar() {
  const { status, data } = useQuery({ queryKey: ['todos'], queryFn: fetchAuthorities })

  const [searchText, setSearchText] = useState('')
  const [recent, setRecent] = useState<Authority[]>([])

  function updateRecent(authority: Authority) {
    // place on head of array, remove duplicates, and limit to 5
    setRecent(
      [authority, ...recent]
      // .filter((item, index, self) => self.findIndex((i) => i.id === item.id) === index)
      // .slice(0, 5)
    )
  }

  function onAuthorityClick(_e: React.MouseEvent<'a', MouseEvent>, _authorityId: number) {
    // Could update recent here
    setSearchText('')
  }

  return (
    <Command
      id='searchbar-label'
      className='rounded-lg border shadow-md'
      label='Search for an authority'
    >
      <CommandInput
        placeholder='Search for an authority...'
        value={searchText}
        onValueChange={(s) => {
          setSearchText(s)
        }}
        data-testid='searchbar-input'
      />
      {searchText.length <= 1 ? (
        <></>
      ) : (
        <CommandList>
          <CommandEmpty data-testid='search-no-results'>No results found.</CommandEmpty>
          <CommandGroup data-testid='search-authorities-group' heading='Authorities'>
            {/* Status error will bubble up to the top level */}
            {status === 'pending' && <CommandEmpty>Loading...</CommandEmpty>}
            {status === 'success' &&
              data.map((authority) => {
                return (
                  <Link
                    data-testid={`result-${authority.name.toLowerCase().replaceAll(' ', '-')}`}
                    key={'link' + authority.id}
                    to='/authority/$authorityId'
                    params={{ authorityId: `${authority.id}` }}
                    onClick={(e) => onAuthorityClick(e, authority.id)}
                    role='link'
                    title={authority.name}
                  >
                    <AuthorityItem
                      key={'item' + authority.id}
                      authority={authority}
                      onClick={updateRecent}
                    ></AuthorityItem>
                  </Link>
                )
              })}
          </CommandGroup>
          <CommandSeparator />
          {/* <CommandGroup heading='Recent'>
            {recent.map((authority) => (
              <AuthorityItem
                key={authority.id}
                authority={authority}
                onClick={updateRecent}
              ></AuthorityItem>
            ))}
          </CommandGroup> */}
        </CommandList>
      )}
    </Command>
  )
}

function AuthorityItem({
  authority,
  onClick
}: {
  authority: Authority
  onClick: (Authority) => void
}) {
  return (
    <CommandItem
      data-testid={`result-${authority.name.toLowerCase().replaceAll(' ', '-')}`}
      value={authority.name}
      key={authority.id}
      onClick={() => onClick(authority)}
    >
      {authority.establishments > 3000 ? (
        <BuildingOffice2Icon className='mr-2 h-4 w-4' />
      ) : authority.establishments > 2000 ? (
        <BuildingOfficeIcon className='mr-2 h-4 w-4' />
      ) : authority.establishments > 1000 ? (
        <HomeModernIcon className='mr-2 h-4 w-4' />
      ) : (
        <HomeIcon className='mr-2 h-4 w-4' />
      )}
      <span>{authority.name}</span>
      {/* <CommandShortcut>âŒ˜1</CommandShortcut> */}
    </CommandItem>
  )
}
