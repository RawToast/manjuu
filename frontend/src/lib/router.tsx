import { Route, Router, rootRouteWithContext } from '@tanstack/react-router'
import { AuthorityInfo, Layout } from '@/pages'
import { QueryClient, queryOptions } from '@tanstack/react-query'
import { fetchAuthorityStats } from '@/lib/client'

export const queryClient = new QueryClient()

const rootRoute = rootRouteWithContext<{
  queryClient: QueryClient
}>()({
  component: Layout
})

export const authoritiesRoute = new Route({
  getParentRoute: () => rootRoute,
  path: 'authority'
})

export const authorityInfoQueryOptions = (authorityId: string) =>
  queryOptions({
    queryKey: ['authorityInfo', { authorityId }],
    queryFn: () => fetchAuthorityStats(authorityId)
  })

export const authorityStats = new Route({
  getParentRoute: () => authoritiesRoute,
  path: '$authorityId',
  component: () => {
    const params = authorityStats.useParams()
    return <AuthorityInfo authorityId={params.authorityId} />
  },
  errorComponent: () => <div>404</div>,
  beforeLoad(opts) {
    opts.context.queryClient.ensureQueryData(authorityInfoQueryOptions(opts.params.authorityId))
  }
})

const authorityRoutes = authoritiesRoute.addChildren([authorityStats])
const routeTree = rootRoute.addChildren([authorityRoutes])

export const router = new Router({
  routeTree,
  context: { queryClient: queryClient }
})

declare module '@tanstack/react-router' {
  interface Register {
    router: typeof router
  }
}
