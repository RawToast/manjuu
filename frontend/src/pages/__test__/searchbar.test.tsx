import { act, render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { expect, test } from 'vitest'

import { SearchBar } from '../searchbar'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { createTestRouter } from '@/lib/test-utils'
import { RouterProvider } from '@tanstack/react-router'

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false
    },
    mutations: {
      retry: false
    }
  }
})

test('Renders searchbar', () => {
  const testRouter = createTestRouter(() => <SearchBar />, queryClient)
  render(
    <QueryClientProvider client={queryClient}>
      <RouterProvider router={testRouter} />,
    </QueryClientProvider>
  )
  const childText = screen.getByPlaceholderText(/Search for an authority/i)
  expect(childText).toBeVisible()
  expect(childText).toMatchSnapshot()
})

test('Highlights best answer in search results', async () => {
  const testRouter = createTestRouter(() => <SearchBar />, queryClient)
  render(
    <QueryClientProvider client={queryClient}>
      <RouterProvider router={testRouter} />,
    </QueryClientProvider>
  )

  const label = await screen.getByLabelText(/Search for an authority/i)
  const text = 'Edi'
  await act(async () => await userEvent.type(label, text))

  expect((label as HTMLInputElement).value).toEqual(text)

  const loading = await screen.findByRole('presentation')
  expect(loading).toBeVisible()

  const option = await screen.findByRole('option', { name: /Fake Edinburgh/i, selected: true })
  expect(option).toBeVisible()
})
