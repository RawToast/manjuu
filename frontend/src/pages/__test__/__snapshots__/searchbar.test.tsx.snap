// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders searchbar 1`] = `
<input
  aria-autocomplete="list"
  aria-controls=":r0:"
  aria-expanded="true"
  aria-labelledby=":r1:"
  autocomplete="off"
  autocorrect="off"
  class="flex h-10 w-full rounded-md bg-transparent py-3 text-sm outline-none placeholder:text-muted-foreground disabled:cursor-not-allowed disabled:opacity-50"
  cmdk-input=""
  data-testid="searchbar-input"
  id=":r2:"
  placeholder="Search for an authority..."
  role="combobox"
  spellcheck="false"
  type="text"
  value=""
/>
`;
